File: windows-client/src/app/services/user/user.service.ts
import { HttpResponse, HttpStatusCode } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { Connection, LoginUser, NewUser, ResetPasswordUser, User, UserAvatar } from '@app/interfaces/user';
import { AvatarService } from '@app/services/avatar/avatar.service';
import { CommunicationSocketService } from '@app/services/communication-socket/communication-socket.service';
import { UserHttpService } from '@app/services/user-http/user-http.service';
import { UserProfile } from '@common/classes/user-profile';
import { UserProfileUI } from '@common/classes/user-profile-ui';
import { UserEvents } from '@common/enums/user.gateway.variables';
import * as crypto from 'crypto-js';
import { Observable, catchError, of, tap, throwError } from 'rxjs';
import { v4 as uuidv4 } from 'uuid';

@Injectable({
    providedIn: 'root',
})
export class UserService {
    loggedInUser: User | null = null;
    tokenKey = 'token';
    usernameColorCache = new Map<string, UserProfileUI>();
    // eslint-disable-next-line max-params
    constructor(
        private userHttpService: UserHttpService,
        private readonly router: Router,
        private socketService: CommunicationSocketService,
        private readonly avatarService: AvatarService,
    ) {
        this.socketService.connect();
    }

    createNewUser(newUser: NewUser): Observable<HttpResponse<string>> {
        this.hashPassword(newUser);
        return this.userHttpService.signUp(newUser).pipe(
            catchError((error) => {
                let errorMessage = 'Une erreur est survenue';
                if (error.status === HttpStatusCode.BadRequest) {
                    errorMessage = "Ce nom d'utilisateur est déjà pris";
                }
                if (error.status === HttpStatusCode.Conflict) {
                    errorMessage = 'Cet email est déjà pris';
                }
                return throwError(() => new Error(errorMessage));
            }),
        );
    }

    login(user: LoginUser): Observable<User> {
        this.hashPassword(user);
        return this.userHttpService.login(user).pipe(
            catchError((error) => {
                let errorMessage = 'Une erreur est survenue';
                if (error.status === HttpStatusCode.Unauthorized) {
                    errorMessage = "Nom d'utilisateur ou mot de passe incorrect";
                } else if (error.status === HttpStatusCode.BadRequest) {
                    errorMessage = 'Utilisateur est déjà connecté';
                }
                return throwError(() => new Error(errorMessage));
            }),
        );
    }

    logout(): void {
        this.avatarService.resetAvatar();
        this.loggedInUser = null;
        this.socketService.send(UserEvents.UserDisconnected);
        this.router.navigate(['/connection']);
    }

    isUserConnected(username: string) {
        this.socketService.send(UserEvents.IsUserConnected, username);
    }

    loginAfterSocket(user: LoginUser): Observable<User> {
        return new Observable((observer) => {
            this.socketService.on(UserEvents.IsUserConnected, (isConnected: boolean) => {
                if (!isConnected) {
                    this.login(user).subscribe({
                        next: (token) => observer.next(token),
                        error: (error) => observer.error(error),
                    });
                } else {
                    observer.error(new Error('Utilisateur est déjà connecté'));
                }
            });
            this.isUserConnected(user.username);
        });
    }

    getCurrentUserId(): string | null {
        return this.loggedInUser ? this.loggedInUser._id : null; // Or this.loggedInUser.username
    }

    isLoggedIn(): boolean {
        const token = localStorage.getItem(this.tokenKey);
        return token != null && token.length > 0;
    }

    getToken(): string | null {
        return this.isLoggedIn() ? localStorage.getItem(this.tokenKey) : null;
    }

    userLoggedIn(): void {
        this.socketService.send(UserEvents.UserConnected, { username: this.loggedInUser?.username, isConnectedToAndroid: false });
    }

    updateConnectionHistory(connection: Connection): void {
        this.userHttpService.updateConnectionHistory(connection).subscribe();
    }

    async setAvatar(data: { [x: string]: string }): Promise<void> {
        return new Promise<void>((resolve, reject) => {
            const avatar: UserAvatar = {
                _id: this.loggedInUser?._id as string,
                avatar: data.isLocalImage ? null : data.avatarData,
                avatarData: data.isLocalImage ? data.avatarData : null,
            };
            this.userHttpService.setAvatar(avatar).subscribe({
                next: (userAvatar) => {
                    this.avatarService.setNetworkAvatar(`${userAvatar.avatar}?${uuidv4()}`);
                    if (this.usernameColorCache.has(this.loggedInUser?.username as string)) {
                        const userProfileUI = this.usernameColorCache.get(this.loggedInUser?.username as string) as UserProfileUI;
                        userProfileUI.avatar = `${userAvatar.avatar}?${uuidv4()}`;
                        this.usernameColorCache.set(this.loggedInUser?.username as string, userProfileUI);
                    }
                    resolve();
                },
                error: (error) => {
                    reject(error);
                },
            });
        });
    }

    updateUsername(newUserName: string): Observable<HttpResponse<string>> {
        const oldKey = this.loggedInUser?.username as string;
        const oldPassword = this.decryptPassword(this.loggedInUser?.password as string, this.loggedInUser?.username as string);
        // TODO: add dto to remove useless email field
        const user = {
            username: newUserName as string,
            email: 'nd@nd.com',
            password: oldPassword as string,
            _id: this.loggedInUser?._id as string,
        };
        this.hashPassword(user);
        return this.userHttpService.updateUser(user).pipe(
            tap(() => {
                this.socketService.send(UserEvents.UserDisconnected);
                localStorage.setItem(this.tokenKey, newUserName);
                if (this.loggedInUser) {
                    this.loggedInUser.username = newUserName;
                    this.loggedInUser.password = user.password;
                }
                this.userLoggedIn();
                if (this.usernameColorCache.has(oldKey)) {
                    const userProfileUI = this.usernameColorCache.get(oldKey) as UserProfileUI;
                    this.usernameColorCache.delete(oldKey);
                    this.usernameColorCache.set(newUserName, userProfileUI);
                }
            }),
            catchError((error) => {
                let errorMessage = 'Une erreur est survenue';
                if (error.status === HttpStatusCode.BadRequest) {
                    errorMessage = "Ce nom d'utilisateur est déjà pris";
                }
                return throwError(() => new Error(errorMessage));
            }),
        );
    }

    // For development purposes only
    deleteAllConnectionHistory(): void {
        this.userHttpService.deleteAllConnectionHistory().subscribe();
    }

    getUserByEmail(email: string): Observable<ResetPasswordUser> {
        return this.userHttpService.getUserByEmail(email).pipe(
            catchError((error) => {
                let errorMessage = 'Une erreur est survenue';
                if (error.status === HttpStatusCode.NotFound) {
                    errorMessage = "Cet email n'est pas associé à un utilisateur";
                }
                return throwError(() => new Error(errorMessage));
            }),
        );
    }

    sendPasswordRecuperationEmail(email: string): Observable<HttpResponse<string>> {
        return this.userHttpService.sendPasswordRecuperationEmail(email).pipe(
            catchError((error) => {
                let errorMessage = 'Une erreur est survenue';
                if (error.status === HttpStatusCode.InternalServerError) {
                    errorMessage = 'Erreur interne du serveur';
                }
                return throwError(() => new Error(errorMessage));
            }),
        );
    }

    verifyCode(code: string): Observable<HttpResponse<string>> {
        return this.userHttpService.verifyCode(code).pipe(
            catchError((error) => {
                let errorMessage = 'Une erreur est survenue';
                if (error.status === HttpStatusCode.BadRequest) {
                    errorMessage = 'Code invalide';
                }
                return throwError(() => new Error(errorMessage));
            }),
        );
    }

    updatePassword(resetPasswordUser: ResetPasswordUser, newPassword: string): Observable<HttpResponse<string>> {
        // TODO: add dto to remove useless email field
        const user = {
            username: resetPasswordUser.username,
            email: 'nd@nd.com',
            password: newPassword,
            _id: resetPasswordUser.userId,
        };
        this.hashPassword(user);
        return this.userHttpService.updatePassword(user).pipe(
            tap(() => {
                if (this.loggedInUser && this.loggedInUser.username === user.username) {
                    this.loggedInUser.password = user.password;
                }
            }),
            catchError((error) => {
                let errorMessage = 'Une erreur est survenue';
                if (error.status === HttpStatusCode.InternalServerError) {
                    errorMessage = 'Erreur interne du serveur';
                }
                return throwError(() => new Error(errorMessage));
            }),
        );
    }

    getUsernameUI(userId: string, username: string): Observable<UserProfileUI> {
        const cachedProfileUI = this.usernameColorCache.get(username);
        if (cachedProfileUI) {
            return of(cachedProfileUI);
        }
        return this.userHttpService.getUsernameUI(userId).pipe(
            tap((result) => {
                this.usernameColorCache.set(username, result);
            }),
            catchError((error) => {
                let errorMessage = 'Une erreur est survenue';
                if (error.status === HttpStatusCode.InternalServerError) {
                    errorMessage = 'Erreur interne du serveur';
                }
                return throwError(() => new Error(errorMessage));
            }),
        );
    }

    getUsernameById(userId: string): Observable<string> {
        return this.userHttpService.getUsernameById(userId).pipe(
            catchError((error) => {
                let errorMessage = 'Une erreur est survenue';
                if (error.status === HttpStatusCode.NotFound) {
                    errorMessage = 'Utilisateur non trouvé';
                }
                return throwError(() => new Error(errorMessage));
            }),
        );
    }

    getUserIdByUsername(username: string): Observable<string> {
        return this.userHttpService.getUserIdByUsername(username).pipe(
            catchError((error) => {
                let errorMessage = 'Une erreur est survenue';
                if (error.status === HttpStatusCode.NotFound) {
                    errorMessage = 'Utilisateur non trouvé';
                }
                return throwError(() => new Error(errorMessage));
            }),
        );
    }

    getUserList(): Observable<UserProfile[]> {
        // eslint-disable-next-line no-underscore-dangle
        return this.userHttpService.getUserList(this.loggedInUser?._id as string).pipe(
            tap(() => {
                catchError((error) => {
                    let errorMessage = 'Une erreur est survenue';
                    if (error.status === HttpStatusCode.InternalServerError) {
                        errorMessage = 'Erreur interne du serveur';
                    }
                    return throwError(() => new Error(errorMessage));
                });
            }),
        );
    }

    getFriendList(userId: string): Observable<string[]> {
        // eslint-disable-next-line no-underscore-dangle
        return this.userHttpService.getFriendList(userId).pipe(
            tap(() => {
                catchError((error) => {
                    let errorMessage = 'Une erreur est survenue';
                    if (error.status === HttpStatusCode.InternalServerError) {
                        errorMessage = 'Erreur interne du serveur';
                    }
                    return throwError(() => new Error(errorMessage));
                });
            }),
        );
    }

    getFriendNotificationCount(userId: string): Observable<number> {
        // eslint-disable-next-line no-underscore-dangle
        return this.userHttpService.getFriendNotificationCount(userId).pipe(
            tap(() => {
                catchError((error) => {
                    let errorMessage = 'Une erreur est survenue';
                    if (error.status === HttpStatusCode.InternalServerError) {
                        errorMessage = 'Erreur interne du serveur';
                    }
                    return throwError(() => new Error(errorMessage));
                });
            }),
        );
    }

    getUsernameByUserId(userId: string): Observable<string> {
        return this.userHttpService.getUsername(userId).pipe(
            catchError((error) => {
                let errorMessage = 'Une erreur est survenue';
                if (error.status === HttpStatusCode.NotFound) {
                    errorMessage = 'Utilisateur non trouvé';
                }
                return throwError(() => new Error(errorMessage));
            }),
        );
    }

    private hashPassword(user: LoginUser | NewUser): void {
        const key = crypto.enc.Utf8.parse(crypto.MD5(user.username).toString());
        const iv = '0102030405060708090a0b0c0d0e0f10';
        user.password = crypto.AES.encrypt(user.password, key, {
            mode: crypto.mode.ECB,
            iv: crypto.enc.Utf8.parse(iv),
        }).toString();
    }

    private decryptPassword(encryptedPassword: string, username: string): string {
        const key = crypto.enc.Utf8.parse(crypto.MD5(username).toString());
        const iv = '0102030405060708090a0b0c0d0e0f10';
        const decrypted = crypto.AES.decrypt(encryptedPassword, key, {
            mode: crypto.mode.ECB,
            iv: crypto.enc.Utf8.parse(iv),
        });
        return decrypted.toString(crypto.enc.Utf8);
    }
}

File: android-client/lib/services/user_service.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import 'package:android_client/classes/connection_type.dart';
import 'package:android_client/classes/user.dart';
import 'package:android_client/components/alert_dialog.dart';
import 'package:android_client/services/avatar_notifier_service.dart';
import 'package:android_client/services/chat_service.dart';
import 'package:android_client/services/socket_service.dart';
import 'package:android_client/services/user_http_service.dart';
import 'package:crypto/crypto.dart' as crypto;
import 'package:encrypt/encrypt.dart' as encrypt;
import 'package:flutter/material.dart';
import 'package:http/http.dart' as http;
import 'package:provider/provider.dart';
import 'package:shared_preferences/shared_preferences.dart';

class UserService {
  final String _tokenKey = 'token';
  static User? loggedInUser;
  late final GlobalKey<NavigatorState> navigatorKey =
      GlobalKey<NavigatorState>();
  late final UserHttpService _userHttpService;
  final _socketService = SocketService();
  static Map<String, UserProfileUI> usernameColorCache =
      <String, UserProfileUI>{};

  UserService() {
    _userHttpService = UserHttpService(http.Client());
    _socketService.connect();
  }

  void createNewUser(NewUser newUser, BuildContext context) async {
    LoginUser user =
        LoginUser(username: newUser.username, password: newUser.password);
    hashPassword(newUser);
    newUser = await _getLocalImageDataFromProvider(context, newUser);
    _userHttpService.signUp(newUser).then((userId) {
      final Connection connection = Connection(
          userId: userId,
          username: newUser.username,
          connectionType: ConnectionType.accountCreation.value,
          connectionTime: DateTime.now().millisecondsSinceEpoch);
      updateConnectionHistory(connection);
      Provider.of<AvatarNotifier>(context, listen: false).resetAvatarProvider();
      loginAfterSocket(
        user,
        context,
      );
    }).catchError((error) {
      openAlertDialog(error.toString(), context);
    });
  }

  void login(LoginUser user, BuildContext context) {
    _userHttpService.login(user).then((token) async {
      token.avatar = UserHttpService.getReplacedUrl(token.avatar);
      SharedPreferences prefs = await SharedPreferences.getInstance();
      prefs.setString(_tokenKey, user.username);
      loggedInUser = token;
      final Connection connection = Connection(
          userId: token.id.toString(),
          username: user.username,
          connectionType: ConnectionType.connection.value,
          connectionTime: DateTime.now().millisecondsSinceEpoch);
      updateConnectionHistory(connection);
      userLoggedIn();
      Navigator.of(context).pushReplacementNamed('/mainScreen');
      Provider.of<AvatarNotifier>(context, listen: false).avatar =
          loggedInUser!.avatar;
    }).catchError((error) {
      openAlertDialog(error.toString(), context);
    });
  }

  void logout(BuildContext context) async {
    _socketService.send('userDisconnected');
    Provider.of<AvatarNotifier>(context, listen: false).resetAvatarProvider();
    Navigator.of(context).pushReplacementNamed('/connectionScreen');
  }

  void isUserConnected(String username) {
    _socketService.send('isUserConnected', username);
  }

  void loginAfterSocket(LoginUser user, BuildContext context) {
    hashPassword(user);
    _socketService.on('isUserConnected', (isConnected) {
      if (isConnected) {
        openAlertDialog('Utilisateur déjà connecté', context);
      } else {
        login(user, context);
      }
      _socketService.off('isUserConnected');
    });
    isUserConnected(user.username);
  }

  Future<void> updateConnectionHistory(Connection connection) async {
    _userHttpService.updateConnectionHistory(connection);
  }

  Future<void> getAvatars(BuildContext context, Function callback) async {
    _userHttpService
        .getPredefinedAvatars()
        .then((data) => {callback(data)})
        .catchError((error) {
      openAlertDialog(error.toString(), context);
      return Future.value(<dynamic>{});
    });
  }

  Future<void> setAvatar(
      BuildContext context, bool isLocalFile, String avatarPath) async {
    UserAvatar userAvatar =
        UserAvatar(id: loggedInUser!.id, avatar: avatarPath);

    if (isLocalFile) {
      userAvatar.avatarData = await _getLocalImageData(userAvatar.avatar);
    }

    try {
      UserAvatar data = await _userHttpService.setAvatar(userAvatar);
      data.avatar = UserHttpService.getReplacedUrl(data.avatar);
      loggedInUser!.avatar = data.avatar;
      if (usernameColorCache.containsKey(loggedInUser!.username)) {
        final userProfileUI =
            usernameColorCache[loggedInUser!.username] as UserProfileUI;
        userProfileUI.avatar = data.avatar;
        usernameColorCache[loggedInUser!.username] = userProfileUI;
      }
    } on Exception catch (error) {
      print(error);
    }
  }

  void userLoggedIn() {
    _socketService.send('userConnected',
        {'username': loggedInUser?.username, 'isConnectedToAndroid': true});
    ChatService.resetSubscribedChannelsAndUnreadMessages();
    ChatService.handleRefreshListOfSubscribedChannels();
  }

  Future<bool> isLoggedIn() async {
    final SharedPreferences prefs = await SharedPreferences.getInstance();
    final String? token = prefs.getString(_tokenKey);
    return token != null && token.isNotEmpty;
  }

  Future<String> getToken() async {
    return await isLoggedIn()
        ? (await SharedPreferences.getInstance()).getString(_tokenKey) ?? ''
        : '';
  }

  void hashPassword(dynamic user) {
    var key = crypto.md5.convert(utf8.encode(user.username)).toString();
    final encrypter = encrypt.Encrypter(
        encrypt.AES(encrypt.Key.fromUtf8(key), mode: encrypt.AESMode.ecb));
    var ivBtyes = Uint8List.fromList(
        [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16]);
    final iv = encrypt.IV(ivBtyes);
    user.password = encrypter.encrypt(user.password, iv: iv).base64;
  }

  Future<Uint8List> _getLocalImageData(avatarUrl) async {
    return await File(avatarUrl).readAsBytes();
  }

  _getLocalImageDataFromProvider(context, newUser) async {
    final avatarNotifier = Provider.of<AvatarNotifier>(context, listen: false);
    if (avatarNotifier.isLocalFile) {
      newUser.avatarData = await _getLocalImageData(avatarNotifier.avatar);
    }
    return newUser;
  }

  String decryptPassword(String encryptedPassword, String username) {
    var key = crypto.md5.convert(utf8.encode(username)).toString();
    final encrypter = encrypt.Encrypter(
        encrypt.AES(encrypt.Key.fromUtf8(key), mode: encrypt.AESMode.ecb));
    var ivBtyes = Uint8List.fromList(
        [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16]);
    final iv = encrypt.IV(ivBtyes);
    return encrypter.decrypt64(encryptedPassword, iv: iv);
  }

  Future<http.Response> updateUsername(String newUsername) async {
    final String oldKey = loggedInUser!.username;
    final SharedPreferences prefs = await SharedPreferences.getInstance();
    var oldPassword =
        decryptPassword(loggedInUser!.password, loggedInUser!.username);
    var user = UpdateUsername(
        username: newUsername,
        email: 'nd@nd.com',
        password: oldPassword,
        id: loggedInUser!.id);
    hashPassword(user);

    final response = _userHttpService.updateUsername(user);
    return response.then((http.Response response) {
      if (response.statusCode != 200) {
        String errorMessage = 'Une erreur est survenue';
        if (response.statusCode == 400) {
          errorMessage = 'Ce nom d\'utilisateur est déjà pris';
        }
        throw Exception(errorMessage);
      }
      _socketService.send('userDisconnected');
      prefs.setString(_tokenKey, newUsername);
      loggedInUser!.username = newUsername;
      loggedInUser!.password = user.password;
      userLoggedIn();
      if (usernameColorCache.containsKey(oldKey)) {
        usernameColorCache[newUsername] =
            usernameColorCache[oldKey] as UserProfileUI;
        usernameColorCache.remove(oldKey);
      }
      return response;
    });
  }

  Future<ResetPasswordUser> getUserByEmail(String email) async {
    try {
      final response = _userHttpService.getUserByEmail(email);
      return response.then((ResetPasswordUser user) {
        return user;
      });
    } on Exception catch (error) {
      throw Exception(error.toString());
    }
  }

  Future<http.Response> sendPasswordRecuperationEmail(String email) {
    final response = _userHttpService.sendPasswordRecuperationEmail(email);
    return response.then((http.Response response) {
      if (response.statusCode != 200) {
        throw Exception('Une erreur est survenue');
      }
      return response;
    }).catchError((error) {
      throw Exception(error.toString());
    });
  }

  Future<http.Response> verifyCode(String code) {
    final response = _userHttpService.verifyCode(code);
    return response.then((http.Response response) {
      if (response.statusCode == 400) {
        throw Exception('Code invalide');
      }
      return response;
    }).catchError((error) {
      throw Exception(error.toString());
    });
  }

  Future<http.Response> updatePassword(
      ResetPasswordUser resetPasswordUser, String password) {
    var user = UpdateUsername(
        username: resetPasswordUser.username,
        email: 'nd@nd.com',
        password: password,
        id: resetPasswordUser.userId);
    hashPassword(user);

    final response = _userHttpService.updatePassword(user);
    return response.then((http.Response response) {
      if (response.statusCode != 200) {
        throw Exception('Une erreur est survenue');
      }
      return response;
    }).catchError((error) {
      throw Exception(error.toString());
    });
  }

  void updateUsernameColor(String usernameColor) {
    UpdateUsernameColor userColor = UpdateUsernameColor(
        userId: loggedInUser!.id, usernameColor: usernameColor);
    _socketService.send('usernameColorUpdated', userColor);
  }

  Future<UserProfileUI> getUsernameUI(String userId, String username) {
    if (usernameColorCache.containsKey(username)) {
      return Future.value(usernameColorCache[username]);
    }
    userId = userId.replaceAll('"', '');
    final response = _userHttpService.getUsernameUI(userId);
    return response.then((UserProfileUI response) {
      response.avatar = UserHttpService.getReplacedUrl(response.avatar);
      response.usernameColor =
          response.usernameColor == '' ? "#000000" : response.usernameColor;
      usernameColorCache[username] = response;
      return response;
    }).catchError((error) {
      throw Exception(error.toString());
    });
  }

  Future<String> getUserIdByUsername(String username) {
    final response = _userHttpService.getUserIdByUsername(username);
    return response.then((String response) {
      return response;
    }).catchError((error) {
      throw Exception(error.toString());
    });
  }

  Future<List<UserProfile>> getUserList() async {
    try {
      return await _userHttpService.getUserList(loggedInUser?.id as String);
    } on Exception catch (error) {
      throw Exception(error.toString());
    }
  }

  Future<List<String>> getFriendList(String userId) async {
    try {
      return await _userHttpService.getFriendList(userId);
    } on Exception catch (error) {
      throw Exception(error.toString());
    }
  }

  Future<String> getFriendNoticationCount(String userId) async {
    try {
      return await _userHttpService.getFriendNotificationCount(userId);
    } on Exception catch (error) {
      throw Exception(error.toString());
    }
  }

  Future<String> getUsernameByUserId(String userId) async {
    final response = _userHttpService.getUsername(userId);
    return response.then((String response) {
      return response;
    }).catchError((error) {
      throw Exception(error.toString());
    });
  }
}
